instruments.go

```diff
@@ -39,54 +39,65 @@ const (
 )
 
 type instruments struct {
-	initOnce        sync.Once
-	initErr         error
 	duration        metric.Int64Histogram
 	requestSize     metric.Int64Histogram
 	responseSize    metric.Int64Histogram
 	requestsPerRPC  metric.Int64Histogram
 	responsesPerRPC metric.Int64Histogram
 }
 
-func (i *instruments) init(meter metric.Meter, isClient bool) {
-	i.initOnce.Do(func() {
-		interceptorType := serverKey
-		if isClient {
-			interceptorType = clientKey
-		}
-		i.duration, i.initErr = meter.Int64Histogram(
-			formatkeys(interceptorType, durationKey),
-			metric.WithUnit(unitMilliseconds),
-		)
-		if i.initErr != nil {
-			return
-		}
-		i.requestSize, i.initErr = meter.Int64Histogram(
-			formatkeys(interceptorType, requestSizeKey),
-			metric.WithUnit(unitBytes),
-		)
-		if i.initErr != nil {
-			return
-		}
-		i.responseSize, i.initErr = meter.Int64Histogram(
-			formatkeys(interceptorType, responseSizeKey),
-			metric.WithUnit(unitBytes),
-		)
-		if i.initErr != nil {
-			return
-		}
-		i.requestsPerRPC, i.initErr = meter.Int64Histogram(
-			formatkeys(interceptorType, requestsPerRPCKey),
-			metric.WithUnit(unitDimensionless),
-		)
-		if i.initErr != nil {
-			return
-		}
-		i.responsesPerRPC, i.initErr = meter.Int64Histogram(
-			formatkeys(interceptorType, responsesPerRPCKey),
-			metric.WithUnit(unitDimensionless),
-		)
-	})
+// makeInstruments creates the metrics for the interceptor. If the metrics
+// cannot be created, an error is returned and the instruments will use the
+// noop metrics.
+func makeInstruments(meter metric.Meter, interceptorType string) (instruments, error) {
+	instruments := instruments{
+		duration:        noop.Int64Histogram{},
+		requestSize:     noop.Int64Histogram{},
+		responseSize:    noop.Int64Histogram{},
+		requestsPerRPC:  noop.Int64Histogram{},
+		responsesPerRPC: noop.Int64Histogram{},
+	}
+	duration, err := meter.Int64Histogram(
+		formatkeys(interceptorType, durationKey),
+		metric.WithUnit(unitMilliseconds),
+	)
+	if err != nil {
```

Comment:
Could multierror to try collect as many valild metrics before failing. This behaviour exits early with previous ones initialised. Didn't think it mattered too much.

Reply:
Hmm, that would be better. Will fix.